From 19255f4b784cedf6dd4bb21e97def348d335dd69 Mon Sep 17 00:00:00 2001
From: Ilija Tovilo <ilija.tovilo@me.com>
Date: Sat, 2 May 2026 01:49:34 +0200
Subject: [PATCH] PHP 5.6: backport OpenSSL 4 compatibility

Backport the ASN1_STRING accessor and error-buffer changes from
https://github.com/php/php-src/commit/19255f4b784cedf6dd4bb21e97def348d335dd69

Backport notes:
Add OpenSSL 4 SSLv3 guards and retain the ASN1 accessor fallback for
older OpenSSL and LibreSSL versions.
Replace the remaining removed ASN1_STRING_data calls.
Replace removed fixed-version TLS methods with generic methods and
explicit minimum and maximum protocol versions.
This branch has no ERR_NUM_ERRORS-based error buffer to change.
---

diff --git a/ext/openssl/openssl.c b/ext/openssl/openssl.c
--- a/ext/openssl/openssl.c
+++ b/ext/openssl/openssl.c
@@ -47,6 +47,10 @@
 #include <openssl/crypto.h>
 #include <openssl/pem.h>
 #include <openssl/err.h>
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || (defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x20700000L)
+#define ASN1_STRING_get0_data ASN1_STRING_data
+#endif
 #include <openssl/conf.h>
 #include <openssl/rand.h>
 #include <openssl/ssl.h>
@@ -667,7 +671,7 @@
 		if (ASN1_STRING_type(str) != V_ASN1_UTF8STRING) {
 			to_add_len = ASN1_STRING_to_UTF8(&to_add, str);
 		} else {
-			to_add = ASN1_STRING_data(str);
+			to_add = ASN1_STRING_get0_data(str);
 			to_add_len = ASN1_STRING_length(str);
 		}
 
@@ -696,7 +700,7 @@
 
 static void add_assoc_asn1_string(zval * val, char * key, ASN1_STRING * str) /* {{{ */
 {
-	add_assoc_stringl(val, key, (char *)str->data, str->length, 1);
+	add_assoc_stringl(val, key, (char *)ASN1_STRING_get0_data(str), ASN1_STRING_length(str), 1);
 }
 /* }}} */
 
@@ -720,22 +724,22 @@
 		return (time_t)-1;
 	}
 
-	if (ASN1_STRING_length(timestr) != strlen((const char*)ASN1_STRING_data(timestr))) {
+	if (ASN1_STRING_length(timestr) != strlen((const char*)ASN1_STRING_get0_data(timestr))) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "illegal length in timestamp");
 		return (time_t)-1;
 	}
 
 	if (ASN1_STRING_length(timestr) < 13) {
-		php_error_docref(NULL TSRMLS_CC, E_WARNING, "unable to parse time string %s correctly", timestr->data);
+		php_error_docref(NULL TSRMLS_CC, E_WARNING, "unable to parse time string %s correctly", ASN1_STRING_get0_data(timestr));
 		return (time_t)-1;
 	}
 
 	if (ASN1_STRING_type(timestr) == V_ASN1_GENERALIZEDTIME && ASN1_STRING_length(timestr) < 15) {
-		php_error_docref(NULL TSRMLS_CC, E_WARNING, "unable to parse time string %s correctly", timestr->data);
+		php_error_docref(NULL TSRMLS_CC, E_WARNING, "unable to parse time string %s correctly", ASN1_STRING_get0_data(timestr));
 		return (time_t)-1;
 	}
 
-	strbuf = estrdup((char *)ASN1_STRING_data(timestr));
+	strbuf = estrdup((char *)ASN1_STRING_get0_data(timestr));
 
 	memset(&thetime, 0, sizeof(thetime));
 
@@ -1264,7 +1268,9 @@
 	}
 
 	php_stream_xport_register("ssl", php_openssl_ssl_socket_factory TSRMLS_CC);
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 	php_stream_xport_register("sslv3", php_openssl_ssl_socket_factory TSRMLS_CC);
+#endif
 #ifndef OPENSSL_NO_SSL2
 	php_stream_xport_register("sslv2", php_openssl_ssl_socket_factory TSRMLS_CC);
 #endif
@@ -1321,7 +1327,9 @@
 #ifndef OPENSSL_NO_SSL2
 	php_stream_xport_unregister("sslv2" TSRMLS_CC);
 #endif
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 	php_stream_xport_unregister("sslv3" TSRMLS_CC);
+#endif
 	php_stream_xport_unregister("tls" TSRMLS_CC);
 	php_stream_xport_unregister("tlsv1.0" TSRMLS_CC);
 #if OPENSSL_VERSION_NUMBER >= 0x10001001L
@@ -1736,7 +1744,7 @@
 		goto cleanup;
 	}
 
-	RETVAL_STRING((char *) ASN1_STRING_data(spki->spkac->challenge), 1);
+	RETVAL_STRING((char *) ASN1_STRING_get0_data(spki->spkac->challenge), 1);
 	goto cleanup;
 
 cleanup:
@@ -1916,8 +1924,8 @@
 	}
 
 	extension_data = X509_EXTENSION_get_data(extension);
-	p = extension_data->data;
-	length = extension_data->length;
+	p = ASN1_STRING_get0_data(extension_data);
+	length = ASN1_STRING_length(extension_data);
 	if (method->it) {
 		names = (GENERAL_NAMES*)(ASN1_item_d2i(NULL, &p, length,
 						       ASN1_ITEM_ptr(method->it)));
@@ -1937,19 +1945,19 @@
 				case GEN_EMAIL:
 					BIO_puts(bio, "email:");
 					as = name->d.rfc822Name;
-					BIO_write(bio, ASN1_STRING_data(as),
+					BIO_write(bio, ASN1_STRING_get0_data(as),
 						  ASN1_STRING_length(as));
 					break;
 				case GEN_DNS:
 					BIO_puts(bio, "DNS:");
 					as = name->d.dNSName;
-					BIO_write(bio, ASN1_STRING_data(as),
+					BIO_write(bio, ASN1_STRING_get0_data(as),
 						  ASN1_STRING_length(as));
 					break;
 				case GEN_URI:
 					BIO_puts(bio, "URI:");
 					as = name->d.uniformResourceIdentifier;
-					BIO_write(bio, ASN1_STRING_data(as),
+					BIO_write(bio, ASN1_STRING_get0_data(as),
 						  ASN1_STRING_length(as));
 					break;
 				default:
diff --git a/ext/openssl/xp_ssl.c b/ext/openssl/xp_ssl.c
--- a/ext/openssl/xp_ssl.c
+++ b/ext/openssl/xp_ssl.c
@@ -35,6 +35,10 @@
 #include <openssl/x509.h>
 #include <openssl/x509v3.h>
 #include <openssl/err.h>
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L || (defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x20700000L)
+#define ASN1_STRING_get0_data ASN1_STRING_data
+#endif
 
 #ifdef PHP_WIN32
 #include "win32/winutil.h"
@@ -402,12 +406,12 @@
 			}
 			OPENSSL_free(cert_name);
 		} else if (san->type == GEN_IPADD) {
-			if (san->d.iPAddress->length == 4) {
+			if (ASN1_STRING_length(san->d.iPAddress) == 4) {
 				sprintf(ipbuffer, "%d.%d.%d.%d",
-					san->d.iPAddress->data[0],
-					san->d.iPAddress->data[1],
-					san->d.iPAddress->data[2],
-					san->d.iPAddress->data[3]
+					ASN1_STRING_get0_data(san->d.iPAddress)[0],
+					ASN1_STRING_get0_data(san->d.iPAddress)[1],
+					ASN1_STRING_get0_data(san->d.iPAddress)[2],
+					ASN1_STRING_get0_data(san->d.iPAddress)[3]
 				);
 				if (strcasecmp(subject_name, (const char*)ipbuffer) == 0) {
 					return 1;
@@ -944,7 +948,7 @@
 				"SSLv2 support is not compiled into the OpenSSL library PHP is linked against");
 		return NULL;
 	} else if (method_value == STREAM_CRYPTO_METHOD_SSLv3) {
-#ifndef OPENSSL_NO_SSL3
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 		return is_client ? SSLv3_client_method() : SSLv3_server_method();
 #else
 		php_error_docref(NULL TSRMLS_CC, E_WARNING,
@@ -952,10 +956,18 @@
 		return NULL;
 #endif
 	} else if (method_value == STREAM_CRYPTO_METHOD_TLSv1_0) {
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+		return is_client ? TLS_client_method() : TLS_server_method();
+#else
 		return is_client ? TLSv1_client_method() : TLSv1_server_method();
+#endif
 	} else if (method_value == STREAM_CRYPTO_METHOD_TLSv1_1) {
 #if OPENSSL_VERSION_NUMBER >= 0x10001001L
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+		return is_client ? TLS_client_method() : TLS_server_method();
+#else
 		return is_client ? TLSv1_1_client_method() : TLSv1_1_server_method();
+#endif
 #else
 		php_error_docref(NULL TSRMLS_CC, E_WARNING,
 				"TLSv1.1 support is not compiled into the OpenSSL library PHP is linked against");
@@ -963,7 +975,11 @@
 #endif
 	} else if (method_value == STREAM_CRYPTO_METHOD_TLSv1_2) {
 #if OPENSSL_VERSION_NUMBER >= 0x10001001L
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+		return is_client ? TLS_client_method() : TLS_server_method();
+#else
 		return is_client ? TLSv1_2_client_method() : TLSv1_2_server_method();
+#endif
 #else
 		php_error_docref(NULL TSRMLS_CC, E_WARNING,
 				"TLSv1.2 support is not compiled into the OpenSSL library PHP is linked against");
@@ -981,7 +997,7 @@
 {
 	long ssl_ctx_options = SSL_OP_ALL;
 
-#ifndef OPENSSL_NO_SSL3
+#if OPENSSL_VERSION_NUMBER < 0x40000000L && !defined(OPENSSL_NO_SSL3)
 	if (!(method_flags & STREAM_CRYPTO_METHOD_SSLv3)) {
 		ssl_ctx_options |= SSL_OP_NO_SSLv3;
 	}
@@ -1485,6 +1501,17 @@
 		return FAILURE;
 	}
 
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L
+	if ((method_flags & (method_flags - 1)) == 0) {
+		int version = method_flags == STREAM_CRYPTO_METHOD_TLSv1_0 ? TLS1_VERSION :
+			(method_flags == STREAM_CRYPTO_METHOD_TLSv1_1 ? TLS1_1_VERSION : TLS1_2_VERSION);
+		if (!SSL_CTX_set_min_proto_version(sslsock->ctx, version) ||
+			!SSL_CTX_set_max_proto_version(sslsock->ctx, version)) {
+			return FAILURE;
+		}
+	}
+#endif
+
 #if OPENSSL_VERSION_NUMBER >= 0x0090806fL
 	if (GET_VER_OPT("no_ticket") && zend_is_true(*val)) {
 		ssl_ctx_options |= SSL_OP_NO_TICKET;
@@ -2419,7 +2446,7 @@
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "SSLv2 support is not compiled into the OpenSSL library PHP is linked against");
 		return NULL;
 	} else if (strncmp(proto, "sslv3", protolen) == 0) {
-#ifdef OPENSSL_NO_SSL3
+#if OPENSSL_VERSION_NUMBER >= 0x40000000L || defined(OPENSSL_NO_SSL3)
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "SSLv3 support is not compiled into the OpenSSL library PHP is linked against");
 		return NULL;
 #else
